src/extension.ts

```diff
@@ -15,15 +15,22 @@ import * as preview from './preview';
 import * as print from './print';
 import * as tableFormatter from './tableFormatter';
 import * as toc from './toc';
+import { setWasm } from './util/slugify';
 
 export function activate(context: ExtensionContext) {
     configNls({ extensionContext: context });
 
-    context.subscriptions.push(
-        configManager, contextServiceManager, decorationManager, commonMarkEngine, mdEngine
-    );
+    // wasm modules need to be imported asynchronously 
+    // we should wait for it to finish to do the rest of the extension
+    import("zola-slug").then((wasm_module) => {
+        setWasm(wasm_module);
```

Comment:
Have you tried to load this on demand in the `slugify.ts`? (considering that this will affect hundreds of thousands of users that do not need this)

Reply:
I'm not very experienced with ts/js and especially not with the async model, I might need help with this. if slugify is called without us having imported the wasm module already then I think it's too late, we would need to the set function to async and every function that calls it? maybe the correct place to import it is when the configuration is loaded or the user switches to this slugification strategy?

Reply:
> we would need to the mark the function `async` and every function that calls it as well and so on?

I also don't have a good solution for the await/async's "contagious" nature. It looks like the call chain is `async generateTocText` (toc.ts line 254) > `getAllTocEntry` > `slugify`. So probably we can just mark the last two functions as async and await the `import` when needed.

Reply:
so the slugify function can't be async because it's part of the MarkdownIt renderrule which expects a synchronous function, the only solution to this I can think of is precomputing slugs and caching them

Reply:
> so the slugify function can't be async because it's part of the MarkdownIt renderrule which expects a synchronous function

Sorry I don't quite understand this. To my memory, the TOC generation doesn't depends on `markdown-it`. Do I forget something?

---

(I tried to debug this on my machine but there was a network issue when installing `wasm-bindgen`. I will try it again later.)

Reply:
makes sense to me

Reply:
I keep running into the problem of contagious async functions, what do you think about just importing it the first time that slugify runs and letting it potentially fail, but it should work the second time?

Reply:
I wish I had time to help. Actually I'm fine with your suggestion (if at least it doesn't affect other users that don't opt-in).

Reply:
ok the current state is the best I can think to write it, I've tried a bunch of other things but the javascript async/await model is like handcuffs, also the syncWebAssembly webpack experiment doesn't work because we would need a async breakoff point somwhere like

import("./util/slugify.ts") instead of import { slugify} from "./util/slugify.ts"

Reply:
since webpack can't include synchronous wasm in the first chunk of the bundle

Reply:
Thanks.

> it's in my opinion the best solution but you stated you didn't want that, what if we did a benchmark how long the async import takes?

If the cost turns out to be neglectable, I won't be against it.

Reply:
actually, i've tested it a few more times now and highest I've gotten was 165ms lowest was 9ms

Reply:
I still think it's fast enough that it's unnoticeable, let me know what you think

Reply:
Then I think it is acceptable. Let's do it. One last request is probably we can add a few comments there so I can easily recall in the future 🤣.